### PR TITLE
Add bookmarks widget and search widget

### PR DESCRIPTION
**Pull request by nyitgroup**
Attached is a new pull request with both widgets as you suggested in #458 

Formatting improvements thanks to Shuhrat

### CODE REVIEW

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java [hunk 1]
```diff
@@ -143,6 +146,8 @@ public boolean replaceRecentRangeWithPage(int deleteRangeStart, int deleteRangeE
   }
 
   public boolean addRecentPage(int page) {
+
+     BookmarksWidget.updateWidget(context);
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java]: This looks wrong to me. What if there are no widgets installed at all?

Comment (nyitgroup) [inline app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java]: I uninstalled the widget and tested it. It didn't cause any problems.

**Comment by nyitgroup** [comment]
I just thought the last PR had too many comments on simple things and mixed up with important comments like the one above.

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java (hunk 1 above)
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java]: > I uninstalled the widget and tested it. It didn't cause any problems.

Perfect engineering mind (:don't touch if it is working:)

**Comment by ozbek** [comment]
> I just thought the last PR had too many comments on simple things and mixed up with important comments like the one above.

Oh well... elsewhere I gave a [brief comment](https://github.com/quran/quran_android/pull/837#issuecomment-291102849), but you wanted a detailed one :)

**Comment by nyitgroup** [comment]
I learned from the detailed comments I even thought about refactoring some of the code but I thought that could wait. Thank you for your attention to the details.

Is there a wiki on the code uml? 
It would make it easier to work with the code models and reactive Java.

File: app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import android.widget.RemoteViews;
+
+import com.quran.labs.androidquran.ui.PagerActivity;
+import com.quran.labs.androidquran.widgets.WidgetService;
+
+public class BookmarksWidget extends AppWidgetProvider {
+
+  @Override
+  public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
+    final int N = appWidgetIds.length;
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.java]: better variable name - ex `numWidgets` or just `widgets` or `count`

File: app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      intent = new Intent(context, SearchActivity.class);
+      pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);
+      widget.setOnClickPendingIntent(R.id.WidgetBtnSearch, pendingIntent);
+      intent = new Intent(context, PagerActivity.class);
+      pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);
+      widget.setOnClickPendingIntent(R.id.WidgetBtnGoToQuran, pendingIntent);
+      widget.setRemoteAdapter(appWidgetIds[i], R.id.listViewWidget,
+          serviceIntent);
+      Intent clickIntent = new Intent(context, PagerActivity.class);
+      PendingIntent clickPI = PendingIntent
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.java]: `clickIntent` or `clickPendingIntent`

File: app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    super.onReceive(context, intent);
+    updateWidget(context);
+
+  }
+
+  public static void updateWidget(Context context) {
+    AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
+    int appWidgetIds[] = appWidgetManager.getAppWidgetIds(new ComponentName(context, BookmarksWidget.class));
+    appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetIds, R.id.listViewWidget);
+
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.java]: remove this extra line

File: app/src/main/java/com/quran/labs/androidquran/SearchActivity.java [hunk 5]
```diff
@@ -53,11 +57,26 @@
...
   public void onCreate(Bundle savedInstanceState) {
     super.onCreate(savedInstanceState);
     setContentView(R.layout.search);
+    searchEditText = (EditText) findViewById(R.id.searchEditText);
+    searchEditText.addTextChangedListener(new TextWatcher() {
+      @Override
+      public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
+      @Override
+      public void onTextChanged(CharSequence s, int start, int before, int count) {
+      if(count>2 )
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/SearchActivity.java]: if (count > 2) {

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+import com.quran.labs.androidquran.R;
+import com.quran.labs.androidquran.dao.Bookmark;
+import com.quran.labs.androidquran.data.QuranInfo;
+import com.quran.labs.androidquran.database.BookmarksDBAdapter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class WidgetListProvider implements RemoteViewsService.RemoteViewsFactory {
+  private ArrayList<ItemList> listItemList = new ArrayList();
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: private List<ItemList> listItemList = new ArrayList<>();

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    this.context = context;
+    appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID,
+        AppWidgetManager.INVALID_APPWIDGET_ID);
+    populateListItem();
+  }
+
+  public void populateListItem() {
+    Context appContext = context.getApplicationContext();
+    BookmarksDBAdapter a = new BookmarksDBAdapter(appContext);
+    List<Bookmark> bookmarksList = a.getBookmarks(0);// 0 = date added
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: isn't there a constant for this - pretty sure there is

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  public long getItemId(int position) {
+    return position;
+  }
+
+  @Override
+  public boolean hasStableIds() {
+    return false;
+  }
+
+  /*
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: fix the comment - you can use intellij's code format feature to format it properly

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  *Similar to getView of Adapter where instead of View
+  *we return RemoteViews
+  *
+  */
+  @Override
+  public RemoteViews getViewAt(int position) {
+    final RemoteViews remoteView = new RemoteViews(
+        context.getPackageName(), R.layout.bookmarks_widget_list_row);
+    ItemList item = listItemList.get(position);
+    remoteView.setTextViewText(R.id.Suratitle, item.sura + "");
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: R.id.sura_title and rename it in the xml

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    remoteView.setTextViewText(R.id.Surametadata, context.getResources().getText(R.string.quran_page) + " " + item.page);
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: R.id.sura_meta_data and rename in xml

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    remoteView.setImageViewResource(R.id.WidgetFavIcon, R.drawable.ic_favorite);
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: R.id.widget_favorite_icon and rename in xml

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    return null;
+  }
+
+  @Override
+  public int getViewTypeCount() {
+    return 1;
+  }
+
+
+  static class BookmarksTable {
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: why do you need this? why not reference the existing one and make it public if needed?

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetService.java [hunk 13]
```diff
@@ -0,0 +1,23 @@
+package com.quran.labs.androidquran.widgets;
+
+import android.appwidget.AppWidgetManager;
+import android.content.Intent;
+import android.widget.RemoteViewsService;
+
+
+public class WidgetService extends RemoteViewsService {
+/*
+* So pretty simple just defining the Adapter of the listview
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetService.java]: fix comment spacing / formatting

File: app/src/main/res/layout/bookmarks_widget.xml [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    android:background="@color/translation_verse_text_color">
+
+  <LinearLayout
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_marginBottom="2dp"
+
```
Comment (ahmedre) [inline app/src/main/res/layout/bookmarks_widget.xml]: remove space

File: app/src/main/res/layout/bookmarks_widget.xml [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+      android:layout_marginTop="2dp"
+      android:background="@drawable/quran_row_background"
+      android:orientation="horizontal">
+
+    <ImageButton
+        android:id="@+id/WidgetIconButton"
```
Comment (ahmedre) [inline app/src/main/res/layout/bookmarks_widget.xml]: lowercase id names using _ to separate words

File: app/src/main/res/values/dimens.xml [hunk 16]
```diff
@@ -36,4 +36,10 @@
   <dimen name="tag_width">50dp</dimen>
   <dimen name="tag_margin">6dp</dimen>
   <dimen name="tag_text_size">10sp</dimen>
+
```
Comment (ahmedre) [inline app/src/main/res/values/dimens.xml]: remove the comment

Comment (nyitgroup) [inline app/src/main/res/values/dimens.xml]: Please, check the latest commits. I tried to change the Icon to text, but it did not look as good.

**Pull request closed without merging**